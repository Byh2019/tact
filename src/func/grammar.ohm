FunC {
  Module = Pragma* Include* ModuleItem*

  //
  // Compiler pragmas and includes
  //

  Pragma = "#pragma" ("allow-post-modification" | "compute-asm-ltr") ";" --literal
         | "#pragma" ("version" | "not-version") versionRange ";" --versionRange
         | "#pragma" "test-version-set" stringLiteral ";" --versionString

  Include = "#include" stringLiteral ";"

  //
  // Top-level, module items
  //

  ModuleItem = GlobalVariablesDeclaration
             | ConstantsDefinition
             | AsmFunctionDefinition
             | FunctionDeclaration
             | FunctionDefinition

  GlobalVariablesDeclaration = "global" NonemptyListOf<GlobalVariableDeclaration, ","> ";"
  GlobalVariableDeclaration = Type? id

  ConstantsDefinition = "const" NonemptyListOf<ConstantDefinition, ","> ";"
  ConstantDefinition = TypeConstant? id "=" Expression

  AsmFunctionDefinition = FunctionCommonPrefix "asm" AsmArrangement? stringLiteral+ ";" --singleLine
                        | FunctionCommonPrefix "asm" AsmArrangement? multiLineString+ ";" --multiLine

  AsmArrangement = "(" "->" integerLiteralDec+ ")" --returns
                 | "(" id+ ~"->" ")" --arguments
                 | "(" id+ "->" integerLiteralDec+ ")" --argumentsToReturns


  FunctionDeclaration = FunctionCommonPrefix ";"

  FunctionDefinition = FunctionCommonPrefix "{" Statement* "}"

  // Common prefix of function declarations and definitions
  FunctionCommonPrefix = Forall? Type functionId Parameters FunctionAttribute*

  // Called "specifiers" in https://docs.ton.org/develop/func/functions#specifiers
  FunctionAttribute = "impure" | "inline_ref" | "inline" | MethodIdValue | "method_id"
  MethodIdValue = "method_id" "(" integerLiteralNonNegative ")"

  //
  // Statements
  //

  Statement = StatementReturn
            | StatementBlock
            | StatementEmpty
            | StatementCondition
            | StatementRepeat
            | StatementUntil
            | StatementWhile
            | StatementTryCatch
            | StatementExpression

  StatementReturn = "return" Expression ";"
  StatementBlock = "{" Statement* "}"
  StatementEmpty = ";"
  StatementCondition = ("if" | "ifnot") Expression "{" Statement* "}" --noElse
                     | ("if" | "ifnot") Expression "{" Statement* "}" "else" "{" Statement* "}" --withElse
                     | ("if" | "ifnot") Expression "{" Statement* "}" ("elseif" | "elseifnot") "{" Statement* "}" --withElseCondition
  StatementRepeat = "repeat" Expression "{" Statement* "}"
  StatementUntil = "do" "{" Statement* "}" "until" Expression ";"
  StatementWhile = "while" Expression "{" Statement* "}"
  StatementTryCatch = "try" "{" Statement* "}" "catch" "(" (unusedId | id) "," (unusedId | id) ")" "{" Statement* "}"
  StatementExpression = Expression ";"

  //
  // Expressions, ordered by precedence (from lowest to highest),
  // with comments referencing exact function names in C++ code of FunC's parser:
  // https://github.com/ton-blockchain/ton/blob/master/crypto/func/parse-func.cpp
  //

  // NOTE: they should parse exactly as FunC 0.4.4 does it, including the issues with precedences of ^ | and similar

  // parse_expr
  // Expression = integerLiteral | stringLiteral | id // FIXME temporary, for tests prior to expressions
  Expression = ExpressionAssign

  // parse_expr10
  ExpressionAssign = ExpressionConditional "="    ExpressionAssign --assign
                   | ExpressionConditional "+="   ExpressionAssign --addAssign
                   | ExpressionConditional "-="   ExpressionAssign --subAssign
                   | ExpressionConditional "*="   ExpressionAssign --mulAssign
                   | ExpressionConditional "/="   ExpressionAssign --divAssign
                   | ExpressionConditional "%="   ExpressionAssign --modAssign
                   | ExpressionConditional "~/="  ExpressionAssign --divRoundAssign
                   | ExpressionConditional "~%="  ExpressionAssign --modRoundAssign
                   | ExpressionConditional "^/="  ExpressionAssign --divCeilAssign
                   | ExpressionConditional "^%="  ExpressionAssign --modCeilAssign
                   | ExpressionConditional "&="   ExpressionAssign --bitwiseAndAssign
                   | ExpressionConditional "|="   ExpressionAssign --bitwiseOrAssign
                   | ExpressionConditional "^="   ExpressionAssign --bitwiseXorAssign
                   | ExpressionConditional "<<="  ExpressionAssign --shlAssign
                   | ExpressionConditional ">>="  ExpressionAssign --shrAssign
                   | ExpressionConditional "~>>=" ExpressionAssign --shrRoundAssign
                   | ExpressionConditional "^>>=" ExpressionAssign --shrCeilAssign
                   | ExpressionConditional

  // parse_expr13
  ExpressionConditional = ExpressionCompare "?" Expression ":" ExpressionConditional --ternary
                        | ExpressionCompare

  // parse_expr15
  ExpressionCompare = ExpressionBitwiseShift "=="  ExpressionBitwiseShift --eq
                    | ExpressionBitwiseShift "<"   ExpressionBitwiseShift --lt
                    | ExpressionBitwiseShift ">"   ExpressionBitwiseShift --gt
                    | ExpressionBitwiseShift "<="  ExpressionBitwiseShift --lte
                    | ExpressionBitwiseShift ">="  ExpressionBitwiseShift --gte
                    | ExpressionBitwiseShift "!="  ExpressionBitwiseShift --neq
                    | ExpressionBitwiseShift "<=>" ExpressionBitwiseShift --spaceship
                    | ExpressionBitwiseShift

  // parse_expr17
  ExpressionBitwiseShift = ExpressionAddBitwise "<<"  ExpressionAddBitwise --shl
                         | ExpressionAddBitwise ">>"  ExpressionAddBitwise --shr
                         | ExpressionAddBitwise "~>>" ExpressionAddBitwise --shrRound
                         | ExpressionAddBitwise "^>>" ExpressionAddBitwise --shrCeil
                         | ExpressionAddBitwise

  // parse_expr20
  ExpressionAddBitwise = "-"? ExpressionMulBitwise "+" ExpressionMulBitwise --add
                       | "-"? ExpressionMulBitwise "-" ExpressionMulBitwise --sub
                       | "-"? ExpressionMulBitwise "|" ExpressionMulBitwise --bitwiseOr
                       | "-"? ExpressionMulBitwise "^" ExpressionMulBitwise --bitwiseXor
                       | "-" ExpressionMulBitwise --minus
                       | ExpressionMulBitwise

  // parse_expr30
  ExpressionMulBitwise = ExpressionUnary "*"  ExpressionUnary --mul
                       | ExpressionUnary "/%" ExpressionUnary --divMod
                       | ExpressionUnary "/"  ExpressionUnary --div
                       | ExpressionUnary "%"  ExpressionUnary --mod
                       | ExpressionUnary "~/" ExpressionUnary --divRound
                       | ExpressionUnary "~%" ExpressionUnary --modRound
                       | ExpressionUnary "^/" ExpressionUnary --divCeil
                       | ExpressionUnary "^%" ExpressionUnary --modCeil
                       | ExpressionUnary "&"  ExpressionUnary --bitwiseAnd
                       | ExpressionUnary

  // parse_expr75
  ExpressionUnary = "~" ExpressionMethodCall --bitwiseNot
                  | ExpressionMethodCall

  // parse_expr80
  ExpressionMethodCall = ExpressionSequence ("." | "~") id ExpressionPrimary --call
                       | ExpressionSequence

  // FIXME: Temporary definition for debugging purposes.
  ExpressionSequence = integerLiteral | multiLineString | stringLiteral | id
  ExpressionPrimary = integerLiteral | multiLineString | stringLiteral | id

  // parse_expr90, FIXME: this is wrong and has to be re-made :)
  // TODO: function call
  // TODO: variable declaration
  // ExpressionSequence = ExpressionPrimary &"(" ExpressionPrimary --call
  //                    | ExpressionPrimary &"[" ExpressionPrimary --access
  //                    | ExpressionPrimary ~("." | "~") id ExpressionPrimary --id
  //                    | ExpressionPrimary

  // parse_expr100, order is important
  // ExpressionPrimary = ExpressionTuple
  //                   | ExpressionTensor
  //                   | integerLiteral
  //                   | stringLiteral
  //                   | hole
  //                   | primitiveType // TODO: re-make this one.
  //                   | id // includes true|false, null|Null
  // TODO: parens!
  // TODO: unit?
  // TODO: slice_string_ident?

  // TODO: ...
  ExpressionTuple = "(" "TODO:" ")"
  ExpressionTensor = "[" "TODO:" "]"

  //
  // Miscellaneous syntactic rules, see: https://ohmjs.org/docs/syntax-reference#syntactic-lexical
  //

  // forall type1, ..., typeN ->
  Forall = "forall" NonemptyListOf<TypeVar, ","> "->"
  TypeVar = "type"? id

  // Parameters, with type holes prohibited during syntax analysis
  Parameters = "(" ListOf<Parameter, ","> ")"
  Parameter = Type id --regular
            | id id --typeVar

  // Arguments -- unused?
  Arguments = "(" ListOf<Argument, ","> ")"
  Argument = id

  // Mapped or unmapped builtin types, where "mapped" refers to:
  // https://docs.ton.org/develop/func/types#functional-type
  Type = TypeBuiltin "->" Type --mapped
       | TypeBuiltin

  // Builtin types
  TypeBuiltin = "int"
                | "cell"
                | "slice"
                | "builder"
                | "cont"
                | "tuple"
                | hole
                | unit
                | Tensor
                | Tuple
                // | id
                // TODO: Use different type signatures in forall functions?

  // Non-functional composite builtin types
  Tensor = "(" NonemptyListOf<Type, ","> ")"
  Tuple = "[" NonemptyListOf<Type, ","> "]"

  // Allowed types of constants
  TypeConstant = "int" | "slice"
  
  //
  // Lexical rules, see: https://ohmjs.org/docs/syntax-reference#syntactic-lexical
  //

  // Special types
  hole = "_" | "var"
  unit = "()"

  // Function identifiers
  functionId = ~("\"" | "{-") ("." | "~")? id

  // Identifiers, with invalidation (keywords, numbers, etc.) during syntax analysis
  // — this makes this parser closer to the C++ one, and also improves error messages

  id = ~("\"" | "{-") (quotedId | plainId)

  quotedId = "`" (~("`" | "\n") any)+ "`"
  plainId = (~(whiteSpace | "(" | ")" | "[ | ]" | "," | "." | ";" | "~") any)+

  // Unused identifiers
  unusedId = "_"

  /*
    FunC can parse much more than Fift can handle. For example, _0x0 and _0 are valid identifiers in FunC, and using either of them compiles and is then interpreted fine by Fift. But if you use both, FunC still compiles, but Fift crashes.

    Same goes for plain identifiers using hashes # or emojis — you can have one FunC function with any of those combinations of characters, but you (generally) cannot have two or more of such functions.
  */

  // Version ranges
  versionRange = ("=" | "^" | "<=" | ">=" | "<" | ">")? integerLiteralDec ("." integerLiteralDec)? ("." integerLiteralDec)?

  // Integers
  integerLiteral = "-"? integerLiteralNonNegative
  integerLiteralNonNegative = integerLiteralHex
                            | integerLiteralDec

  // hexDigit is defined in Ohm's built-in rules as: hexDigit = "0".."9" | "a".."f" | "A".."F"
  integerLiteralHex = "0x" hexDigit+

  // digit is defined in Ohm's built-in rules as: digit = "0".."9"
  integerLiteralDec = digit+

  // Strings
  stringLiteral = "\"" (~"\"" any)* "\"" stringType?
  stringType = "s" --sliceHex
             | "a" --sliceInternalAddress
             | "u" --intHex
             | "h" --int32Sha256
             | "H" --int256Sha256
             | "c" --intCrc32
  multiLineString = "\"\"\"" (~"\"\"\"" any)* "\"\"\""

  // Whitespace
  space += comment | lineTerminator
  whiteSpace = "\t" | " " | lineTerminator

  // Comments
  comment = ";;" (~lineTerminator any)* --singleLine
          | multiLineComment --multiLine

  // Nested multi-line comments
  multiLineComment = "{-" (~("-}" | "{-") any)* multiLineComment? (~"-}" any)* "-}"

  lineTerminator = "\n" | "\r" | "\u2028" | "\u2029"
}
