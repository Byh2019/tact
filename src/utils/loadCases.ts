import fs from "fs";

/**
 * @param src Path to the directory with files
 * @param ext Optional extension of the file, without the dot prefix
 */
export function loadCases(src: string, ext?: string) {
    const recs = fs.readdirSync(src);
    const res: { name: string; code: string }[] = [];
    const extWithDot = `.${ext ? ext.toLowerCase() : "tact"}`;

    for (const r of recs) {
        if (r.endsWith(extWithDot)) {
            res.push({
                name: r.slice(0, r.length - extWithDot.length),
                code: fs.readFileSync(src + r, "utf8"),
            });
        }
    }
    res.sort((a, b) => a.name.localeCompare(b.name));
    return res;
}
