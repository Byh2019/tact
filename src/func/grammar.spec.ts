import { FuncParseError, FuncSyntaxError, match, parseFile } from "./grammar";
import { loadCases } from "../utils/loadCases";

describe("FunC grammar and parser", () => {
    beforeEach(() => {});
    const ext = "fc";
    let matchedAll = true;

    // Checking that valid FunC files match the grammar
    for (const r of loadCases(__dirname + "/grammar-test/", ext)) {
        it(r.name + " should match the grammar", () => {
            const res = match(r.code);

            if (res.ok === false) {
                matchedAll = false;
                console.log(res.message, res.interval.getLineAndColumn());
            }

            expect(res.ok).toStrictEqual(true);
        });
    }

    // If didn't match the grammar, don't throw any more errors from full parse
    if (!matchedAll) {
        return;
    }

    // Checking that valid FunC files parse
    for (const r of loadCases(__dirname + "/grammar-test/", ext)) {
        it("should parse " + r.name, () => {
            let parsed: Object | undefined;
            try {
                parsed = parseFile(r.code, r.name + `.${ext}`);
            } finally {
                expect(parsed).not.toBe(undefined);
            }
        });
    }

    // Checking that invalid FunC files does NOT parse
    for (const r of loadCases(__dirname + "/grammar-test-failed/", ext)) {
        it("should NOT parse " + r.name, () => {
            expect(() =>
                parseFile(r.code, r.name + `.${ext}`),
            ).toThrowErrorMatchingSnapshot();
        });
    }
});
