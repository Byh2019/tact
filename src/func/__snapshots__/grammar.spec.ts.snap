// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FunC grammar and parser should NOT parse id-arith-operator 1`] = `"semantics(...).astOfModule is not a function"`;

exports[`FunC grammar and parser should NOT parse id-assign-operator 1`] = `"semantics(...).astOfModule is not a function"`;

exports[`FunC grammar and parser should NOT parse id-bitwise-operator 1`] = `
"id-bitwise-operator.fc:1:5: Parse error: expected not (a whiteSpace or "(" or ")" or "[ | ]" or "," or "." or ";" or "~") or "\`"

Line 1, col 5:
> 1 | () ~();
          ^
  2 | 
"
`;

exports[`FunC grammar and parser should NOT parse id-comma 1`] = `
"id-comma.fc:1:16: Parse error: expected "("

Line 1, col 16:
> 1 | () send_message,then_terminate();
                     ^
  2 | 
"
`;

exports[`FunC grammar and parser should NOT parse id-comparison-operator 1`] = `"semantics(...).astOfModule is not a function"`;

exports[`FunC grammar and parser should NOT parse id-control-keyword 1`] = `"semantics(...).astOfModule is not a function"`;

exports[`FunC grammar and parser should NOT parse id-delimiter 1`] = `"semantics(...).astOfModule is not a function"`;

exports[`FunC grammar and parser should NOT parse id-directive 1`] = `"semantics(...).astOfModule is not a function"`;

exports[`FunC grammar and parser should NOT parse id-dot 1`] = `
"id-dot.fc:1:7: Parse error: expected "("

Line 1, col 7:
> 1 | () msg.sender();
            ^
  2 | 
"
`;

exports[`FunC grammar and parser should NOT parse id-keyword 1`] = `"semantics(...).astOfModule is not a function"`;

exports[`FunC grammar and parser should NOT parse id-multiline-comments 1`] = `
"id-multiline-comments.fc:1:11: Parse error: expected not (a whiteSpace or "(" or ")" or "[ | ]" or "," or "." or ";" or "~"), "\`", or "->"

Line 1, col 11:
> 1 | () {-aaa-}();
                ^
  2 | 
"
`;

exports[`FunC grammar and parser should NOT parse id-number 1`] = `"semantics(...).astOfModule is not a function"`;

exports[`FunC grammar and parser should NOT parse id-number-decimal 1`] = `"semantics(...).astOfModule is not a function"`;

exports[`FunC grammar and parser should NOT parse id-number-hexadecimal 1`] = `"semantics(...).astOfModule is not a function"`;

exports[`FunC grammar and parser should NOT parse id-number-hexadecimal-2 1`] = `"semantics(...).astOfModule is not a function"`;

exports[`FunC grammar and parser should NOT parse id-number-neg-decimal 1`] = `
"id-number-neg-decimal.fc:2:1: Parse error: expected end of input, "[", "(", "()", "var", "_", "tuple", "cont", "builder", "slice", "cell", "int", "const", or "global"

Line 2, col 1:
  1 | () -1();
> 2 | native idTest();
      ^
  3 | 
"
`;

exports[`FunC grammar and parser should NOT parse id-number-neg-hexadecimal 1`] = `"semantics(...).astOfModule is not a function"`;

exports[`FunC grammar and parser should NOT parse id-only-underscore 1`] = `"semantics(...).astOfModule is not a function"`;

exports[`FunC grammar and parser should NOT parse id-parens 1`] = `
"id-parens.fc:1:14: Parse error: expected not (a whiteSpace or "(" or ")" or "[ | ]" or "," or "." or ";" or "~") or "\`"

Line 1, col 14:
> 1 | () take(first)Entry();
                   ^
  2 | 
"
`;

exports[`FunC grammar and parser should NOT parse id-semicolons 1`] = `
"id-semicolons.fc:2:1: Parse error: expected "("

Line 2, col 1:
  1 | () pa;;in"\`aaa\`"();
> 2 | 
      ^
"
`;

exports[`FunC grammar and parser should NOT parse id-space 1`] = `
"id-space.fc:1:8: Parse error: expected "("

Line 1, col 8:
> 1 | () foo foo();
             ^
  2 | 
"
`;

exports[`FunC grammar and parser should NOT parse id-string 1`] = `
"id-string.fc:1:4: Parse error: expected not ("\\"" or "{-") or "->"

Line 1, col 4:
> 1 | () "not_a_string();
         ^
  2 | 
"
`;

exports[`FunC grammar and parser should NOT parse id-type-keyword 1`] = `
"id-type-keyword.fc:1:8: Parse error: expected not (a whiteSpace or "(" or ")" or "[ | ]" or "," or "." or ";" or "~") or "\`"

Line 1, col 8:
> 1 | () ->();
             ^
  2 | 
"
`;

exports[`FunC grammar and parser should NOT parse id-unclosed-parens 1`] = `
"id-unclosed-parens.fc:1:9: Parse error: expected not (a whiteSpace or "(" or ")" or "[ | ]" or "," or "." or ";" or "~") or "\`"

Line 1, col 9:
> 1 | () aa(bb();
              ^
  2 | 
"
`;
